rubix-core/src/main/java/com/qubole/rubix/core/CachedReadRequestChain.java

```diff
@@ -74,44 +86,140 @@ public Integer call()
     }
 
     checkState(isLocked, "Trying to execute Chain without locking");
-    for (ReadRequest readRequest : readRequests) {
-      if (cancelled) {
-        propagateCancel(this.getClass().getName());
-      }
-      int nread = 0;
-      int leftToRead = readRequest.getActualReadLength();
-      log.debug(String.format("Processing readrequest %d-%d, length %d", readRequest.actualReadStart, readRequest.actualReadEnd, leftToRead));
-      while (nread < readRequest.getActualReadLength()) {
-        int readInThisCycle = Math.min(leftToRead, directBuffer.capacity());
-        directBuffer.clear();
-        int nbytes = fileChannel.read(directBuffer, readRequest.getActualReadStart() + nread);
-        if (nbytes <= 0) {
-          break;
+
+    RandomAccessFile raf = null;
+    FileInputStream fis = null;
+    FileChannel fileChannel = null;
+    int readRequestIndex = 0;
+    boolean needsInvalidation = false;
+
+    try {
+      raf = new RandomAccessFile(localCachedFile, "r");
+      fis = new FileInputStream(raf.getFD());
+      fileChannel = fis.getChannel();
+
+      for (ReadRequest readRequest : readRequests) {
+        if (cancelled) {
+          propagateCancel(this.getClass().getName());
+        }
+        int nread = 0;
+        int leftToRead = readRequest.getActualReadLength();
+        log.debug(String.format("Processing readrequest %d-%d, length %d", readRequest.actualReadStart, readRequest.actualReadEnd, leftToRead));
+        while (nread < readRequest.getActualReadLength()) {
+          int readInThisCycle = Math.min(leftToRead, directBuffer.capacity());
+          directBuffer.clear();
+          int nbytes = fileChannel.read(directBuffer, readRequest.getActualReadStart() + nread);
+          if (nbytes <= 0) {
+            break;
+          }
+          directBuffer.flip();
+          int transferBytes = Math.min(readInThisCycle, nbytes);
+          directBuffer.get(readRequest.getDestBuffer(), readRequest.getDestBufferOffset() + nread, transferBytes);
+          leftToRead -= transferBytes;
+          nread += transferBytes;
+        }
+        log.debug(String.format("CachedFileRead copied data [%d - %d] at buffer offset %d",
+                readRequest.getActualReadStart(),
+                readRequest.getActualReadStart() + nread,
+                readRequest.getDestBufferOffset()));
+
+        if (nread != readRequest.getActualReadLength()) {
+          needsInvalidation = true;
+          log.error(String.format("Cached read length didn't match with requested read length for file %s. " +
+                  " Falling back reading from object store.", localCachedFile));
+          directDataRead = readFromRemoteFileSystem(readRequests.indexOf(readRequest));
```

Comment:
why trust any data in that file? I suggest we just invalidate and return data from direct reads for all RRs in the chain

Reply:
Done. Reading the entire request from remote object store.